### ISSUE TITLE
ClassDef checker error after linking when only using the class data of a non-native JS class

### ISSUE THREAD

**Issue by futile** [scala-js/scala-js#4850]
In a project (non-public :( ) that uses [tapir](https://github.com/softwaremill/tapir), compilation with 1.13.0 worked fine, while the upgrade to 1.13.1 throws the following error:

```
[error] file:/home/runner/work/tapir/tapir/serverless/aws/lambda/src/main/scalajs/sttp/tapir/serverless/aws/lambda/js/AwsJsRequest.scala(29:7:ClassDef): JS classes and module classes must have a constructor
[error] java.util.concurrent.ExecutionException: Boxed Error
[error] 	at scala.concurrent.impl.Promise$.resolver(Promise.scala:87)
[error] 	at scala.concurrent.impl.Promise$.scala$concurrent$impl$Promise$$resolveTry(Promise.scala:79)
[error] 	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:284)
[error] 	at scala.concurrent.Promise.complete(Promise.scala:53)
[error] 	at scala.concurrent.Promise.complete$(Promise.scala:52)
[error] 	at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:187)
[error] 	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
[error] 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
[error] 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
[error] 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
[error] 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
[error] 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
[error] 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
[error] 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
[error] Caused by: java.lang.AssertionError: There were 1 ClassDef checking errors after optimizing. Please report this as a bug.
[error] 	at org.scalajs.linker.frontend.Refiner$ClassInfoCache.update(Refiner.scala:161)
[error] 	at org.scalajs.linker.frontend.Refiner$ClassInfoCache.$anonfun$loadInfo$2(Refiner.scala:147)
[error] 	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
[error] 	at scala.util.Success.$anonfun$map$1(Try.scala:255)
[error] 	at scala.util.Success.map(Try.scala:213)
[error] 	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
[error] 	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
[error] 	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
[error] 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
[error] 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
[error] 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
[error] 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
[error] 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
[error] 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
[error] 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

The offending line seems to be this one:
https://github.com/softwaremill/tapir/blob/master/serverless/aws/lambda/src/main/scalajs/sttp/tapir/serverless/aws/lambda/js/AwsJsRequest.scala#L29

No idea how to minimize this, but since it's a public library, it should be possible to reproduce. This bug occurs when compiling tapir 1.2.12, which is also the most recent release, released last week (at the time of writing).

Please let me know if/what further information can help, or to test a possible fix! Really looking forward to those incremental-comp performance upgrades in 1.13.1 :) Thanks!

**Comment by sjrd** [comment]
Hum, interesting. FWIW, the bug probably existed before 1.13.1, but did not do ClassDef checks after the optimizer before.

You should be able to work around the issue by disabling IR checks (`.withCheckIR(false)` on `scalaJSLinkerConfig`). IIIUC it's the default, so you should already have a `withCheckIR` somewhere in your build that makes it `true`.

**Comment by sjrd** [comment]
Alternative workaround: add a
```scala
scala.scalajs.js.constructorOf[sttp.tapir.serverless.aws.lambda.js.AwsJsRequestContext]
```
somewhere in your code (not in dead code), such as in your `main` method. That should force the reachability analysis to preserve the constructor of that class.

**Comment by futile** [comment]
Thanks for the quick reply!

From a quick grep for `withCheckIR` through the source of my project, as well as tapir itself, nothing came up. I took a look at where `withCheckIR` is used in scalajs, and found this line:

https://github.com/scala-js/scala-js/blob/cb71886c5bf5534fdc2f232f6c7bd552561d8c36/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala#L474

It seems that the IR-checks are enabled for `fullOptJS` by default? And this assertion happens in my project when building `fullOptJS`, so that would match up.

I'll test out the workaround and report back as well.

**Comment by futile** [comment]
I can confirm two things already:

* The assertion only occurs when compiling `fullOptJS`, while `fastOptJS` compiles fine.
* Using the `constructorOf[...]` workaround, the build works again, so the workaround works

I didn't manage to find how I enable `checkIR` through sbt. `.settings(fullOptJS / checkIR := false)` didn't work, maybe you can tell me what the easiest way to enable `checkIR` is?

**Comment by sjrd** [comment]
Ah yes, it's true by default in fullLink. I had forgotten about that.

The correct setting is:
```scala
.settings(Compile / fullLinkJS / scalaJSLinkerConfig ~= { _.withCheckIR(false) }
```

**Comment by sjrd** [comment]
It's unlikely that we'll make a quick release for this, unless a number of other people experience the same issue. We normally stick to our 2-month schedule.

**Comment by futile** [comment]
Alright, thanks for letting me know, and for the quick replies again!

**Comment by sjrd** [comment]
It turns out this was already fixed in ae17b935d1cd388e67c58744d69484d25e4ff78c. Consolidating is good. :p PR to enshrine it in tests open at #4852.
